Use PJ in `oij` ##json

officialcjunior opened the pull request:
<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

`oij` currently works by constructing a JSON string and printing it. Here, I've reimplemented it using the PJ API.

**Test plan**
Using this, `oij` on one of the test binary prints:
`[{"raised":false,"fd":3,"uri":"bins/elf/crackme0x05","from":0,"writable":false,"size":7656},{"raised":true,"fd":5,"uri":"bins/elf/crackme0x00b","from":0,"writable":false,"size":7375}]`

**Closing issues**

None

karliss commented:
iAj or oj?

karliss commented:
Please add a test as asked in the checklist. If you did that, you would see that the code isn't working as you described.

trufae commented on libr/core/cfile.c:
```diff
@@ -1072,15 +1072,22 @@ R_API int r_core_file_list(RCore *core, int mode) {
+			pj_ki (pj, "size", (int) r_io_desc_size (desc));
+			pj_end (pj);
+			pj_end (pj);
+			r_cons_printf ("%s\n", pj_string (pj));
```
Use r_cons_println

officialcjunior replied:
Replaced with r_cons_println.
Resolving this as marked

trufae commented on libr/core/cfile.c:
```diff
@@ -1072,15 +1072,22 @@ R_API int r_core_file_list(RCore *core, int mode) {
+		case 'j': {  // "oj"
+			PJ * pj = pj_new ();
+			pj_o (pj);
+			pj_a (pj);
```
an array inside an object? this doesnt seems good json to me. i think pj api should check for those invalid json serialization problems

officialcjunior replied:
Oh, okay.
I just went with that, as that's what I've seen commonly implemented, and to make it similar to the current output.

What should be changed?

trufae replied:
{[1,2,3]} is invalid json. Just remove the pj_a

karliss commented:
Still wrong, still no test. And that part of code isn't even executed by `oj`.

officialcjunior commented:
Okay. I had mistook this as the code of `oj` as it has similar outputs. I still have no idea what this command is. I realize that I am very short-sighted and senseless, at the moment. If you could tell me a way to find what this command is, I can rename this PR and add a test for it.

karliss commented:
When writing code and doing local testing you should notice that making changes to code have no effect on output. A dumb but effective way for verifying that you are working with the right code is adding extra print statements, if their output is not visible then something is wrong.

As with all other commands there two simple approaches that work most of the time for locating required code. One is to search by command name something like "oj". I assume this is what you did, but it didn't work out this time. Just searching for such comment comment doesn't guarantee that it's correct code so it's worth verifying the function containing it and where it's called.
If you look at search results carefully you can see that there are two results, this is a good indicator you have to start paying attention and reading the code more carefully. This can be considered bottom to top approach. 

Other approach is following the path from top to bottom. Start with the top level command handler starting with cmd_open and follow the the switch statements and function calls handling command letters one by one until you reach `oj`. In case of `oj` that's only two steps.

How could you know this being new to r2 source code Follow the function calls. You can apply this approach to learning any code base.
* Get your bearing. Locate code handling the part you are interested in or something similar. Good approach for this searching for ui strings, help messages, warnings that appear when interacting with corresponding feature. Especially for command line tools like r2.
* Follow the function calls and other references up. Sometimes the string you found might not be directly next the function you want or if you started with some similar method you might need to go a few levels up. Follow the function calls and variable names. Until you reach the common code. While doing this also pay attention of code structure. This will help in next step when you follow the code in opposite direction.
* Follow the function calls top to bottom. Use the things you learned previously for which path is more likely.

IDE features for finding function declaration or references to function are nice but even with regular you can usually find the right thing easily as global function names are typically sufficiently long and unique. Most code editors have a feature for searching text in all files inside project directory. Learn to use it. In case of less integrated development process you can use tools like grep.

Remember that software development isn't just writing code but also reading and navigating code written by others. And you need to practice all of those skills.

In this specific case it seems that the code you modified used to be used by `oj` command but not anymore so it should be cleaned up so that there aren't two implementations.

officialcjunior commented:
Thank you so much, @karliss.
And I did find it. It's `oij`.

Also, about removing `pj_a()`:
Currently it r_cons_prints the square brackets at line [L1064](
  https://github.com/radareorg/radare2/blob/c06d82eb9b3ba75f29940a7550af0ed68db8f66b/libr/core/cfile.c#L1064) and [L1147](https://github.com/radareorg/radare2/blob/c06d82eb9b3ba75f29940a7550af0ed68db8f66b/libr/core/cfile.c#L1147).

Removing the `pj_a` and `r_cons_println`-ing the `pj_string` will output `oij` like this: (as there's a newline) 
```
[{"raised":"true","fd":3,"uri":"../ch18.bin","from":0,"writable":"false","size":5240}
]
```

So, I think it's better to remove printing square brackets and get a `pj_a()`, to fix this.

What should I do?

karliss commented:
> What should I do?

1) Learn how to run testsuite, see README in test folder
2) Create a test, doesn't need to be final, you can adjust it later. Try adding a bit more than just basic case, for example add a case where there are two files in the list.
3) Run the test
4) Observe that for test with two opened files it doesn't properly print them
5) Get rid of manual `[` printing and use appropriate PJ api for describing list of elements as you suggested yourself. The whole point of this PR was to replace manual JSON printing with use of PJ. Using PJ for parts of single JSON output and then gluing them together like it's done currently defeats the point of this PR.

officialcjunior commented:
Removed the manual printing part in the output.

I also added a basic test and ran it. I'm not sure about whether I have done it correctly and whether this is what you meant by adding two files as, I didn't encounter any failed tests.
 
```
$ r2r -i db/cmd/cmd_open 
Running from /home/cjunior/Downloads/programs/radare2/test
Loaded 30 tests.
Skipping json tests because jq is not available.
[30/30]                    29 OK         1 BR        0 XX        0 FX
Finished in 1 seconds.

```

officialcjunior commented:
Removed the 'manual printing' part

Added a basic test for this. I'm not sure whether I've done it right, as:
1. I'm slightly confused whether this is what you meant by opening two files.
2. The test didn't encounter any errors:

```
$ r2r -i db/cmd/cmd_open 
Running from /home/cjunior/Downloads/programs/radare2/test
Loaded 30 tests.
Skipping json tests because jq is not available.
[30/30]                    29 OK         1 BR        0 XX        0 FX
Finished in 0 seconds.

```

karliss commented:
> I'm slightly confused whether this is what you meant by opening two files.

Two files at the same time so that the list contains more than one entry instead of two files. Two files one at a time doesn't add any value to the test. See `o?` for commands that can be used to open additional files.

officialcjunior commented:
Under, `o?`, I tried `o [filename]`. It doesn't accept two arguments, so I opened two files twice, but that seems to have the same effect as opening two files normally.

Then, there's `oc`, `of` and `oi`, which I'm not sure what it is used for.

Please share the syntax to open two files the way you mean,

karliss commented:
Looks like it the same.Not sure if it's an intended way of using r2r but does what I wanted and demonstrates that output isn't valid json.

officialcjunior commented:
```
[0x08048320]> oj
[{"raised":true,"fd":3,"uri":"ch18.bin","from":0,"writable":false,"size":5240},{"raised":false,"fd":4,"uri":"null://8","from":0,"writable":true,"size":8},{"raised":false,"fd":5,"uri":"malloc://983040","from":0,"writable":true,"size":983040}]
[0x08048320]> oij
[{"raised":"true","fd":3,"uri":"ch18.bin","from":0,"writable":"false","size":5240}]
```
So, the JSON notation is correct now, isn't it?

karliss commented:
Take a look at expected output in the test, that isn't valid JSON.

officialcjunior commented:
I hope that's fixed now, if not, please request changes.

Maijin commented:
Add a test in JSON test db/json/

officialcjunior commented:
Yeah, is that it @Maijin ? Just adding the word?

Maijin commented:
yes

trufae commented on libr/core/cfile.c:
```diff
@@ -1072,15 +1069,21 @@ R_API int r_core_file_list(RCore *core, int mode) {
+			pj_ki (pj, "fd", (int) f->fd);
+			pj_ks (pj, "uri", desc->uri);
+			pj_kn (pj, "from", (ut64) from);
+			pj_ks (pj, "writable", r_str_bool (desc->perm & R_PERM_W));
```
writable should be a bool. use pj_kb, not a string

officialcjunior replied:
Replaced with `pj_kb`

trufae commented on libr/core/cfile.c:
```diff
@@ -1072,15 +1069,21 @@ R_API int r_core_file_list(RCore *core, int mode) {
+		case 'j': {  // "oij"
+			PJ * pj = pj_new ();
+			pj_o (pj);
+			pj_ks (pj, "raised", r_str_bool (core->io->desc->fd == f->fd));
```
raised must be a bool. not a string

trufae replied:
ping

officialcjunior replied:
Replaced with `pj_kb`

trufae commented on libr/core/cfile.c:
```diff
@@ -1072,15 +1069,21 @@ R_API int r_core_file_list(RCore *core, int mode) {
-				(int) r_io_desc_size (desc),
-				iter->n? ",": "");
+		case 'j': {  // "oij"
+			PJ * pj = pj_new ();
```
check for null

officialcjunior replied:
Did you mean:
```
if (!pj) {
  free (out);
  break;
}
```

trufae replied:
yes

officialcjunior commented:
So, I did what you asked me to, but the bool for `writable` got swapped, as you can see in the tests.
Is that alright?

karliss commented:
Output still isn't a valid JSON. `{}{}` is not JSON.

trufae reviewed (changes requested):
Move the pj_new outside the loop and you are opening one object and closing it twice. You should open an array outside the loop and an object inside...

officialcjunior commented:
I think I've correctly did what you had asked. Now, the output is :

`[{"raised":true,"fd":3,"uri":"test/bins/elf/crackme0x05","from":0,"writable":true,"size":7656}`

But as @karliss had pointed out earlier, this fails when we open two files.

officialcjunior commented:
Hey @XVilka, this seems to be done.

Maijin commented on test/db/cmd/cmd_open:
```diff
@@ -105,8 +105,7 @@ bins/elf/crackme0x00b
 EXPECT=<<EOF
-{"raised":true,"fd":3,"uri":"bins/elf/crackme0x05","from":0,"writable":true,"size":7656}
-{"raised":true,"fd":5,"uri":"bins/elf/crackme0x00b","from":0,"writable":true,"size":7375}
+[{"raised":false,"fd":3,"uri":"bins/elf/crackme0x05","from":0,"writable":false,"size":7656},{"raised":true,"fd":5,"uri":"bins/elf/crackme0x00b","from":0,"writable":false,"size":7375}]
```
don't forget to also add a json test in https://github.com/radareorg/radare2/tree/master/test/db/json

officialcjunior replied:
Hey, I have added that:

https://github.com/radareorg/radare2/pull/17642/files#diff-11f08680e054bf1235c7c156ec3962f4272ae558b64c44158f965f1d287cbf49

officialcjunior replied:
Marking as resolved.

eagleoflqj commented:
Ping @XVilka

XVilka commented:
Sorry for the delay, didn't notice it before.
